dev tools quick-install script: Paul's changes (#373)
holochain/docs-pages: src/pages/get-building/install-advanced.md

---

::: coreconcepts-intro
This guide assumes that you've already followed the [quick installation guide](../../install/) and want to learn more about the setup. It describes how to manually recreate and maintain the development environment, use your default shell and preferred code editor with Nix, or install specific versions of Holochain, and explains why we use `nix develop` in the first place.
:::

## Holonix, the Holochain app development environment

Each Holochain application repository will contain its own setup of the development environment.
If you use the scaffolding to generate the project structure, this will already be taken care of in the scaffolded directory.

If you want to learn more about how this setup works and how to create it manually and how to maintain it, please find all the information below.

> ### Holonix's usage of [Nix's Flakes](https://nixos.wiki/wiki/Flakes) features

Reviewer: ```suggestion
### Holonix's usage of [Nix Flakes](https://nixos.wiki/wiki/Flakes) features
```

In principle a flake implements a function that produces a set of _outputs_ from a given set of _inputs_, keeping the side-effects to an absolute minimum if not at zero.

#### `inputs`

> This flake declares one input named `holochain-flake` that the Holochain Github repository. This input will look for a `flake.nix` in the default branch of the remote repository. The `versions` input of the `holochain-flake` input is explicitly specified to track the _0_1_ series, which refers to Holochain's Beta 0.1 and its compatible tools.

Reviewer: ```suggestion
This flake declares one input named `holochain-flake` that represents the Holochain Github repository. This input will look for a `flake.nix` in the default branch of the remote repository. The `versions` input of the `holochain-flake` input is explicitly specified to track the _0_1_ series, which refers to Holochain's Beta 0.1 and its compatible tools.
```

> The flake follows (think _inherits_) the `nixpkgs` input of the `holochain-flake` input. This ensures that your development environment passes all the same buildinputs to the component packages, giving you very high chances to make use of our Cachix binary cache.

Reviewer: ```suggestion
The flake follows (think _inherits_) the `nixpkgs` input of the `holochain-flake` input. This ensures that your development environment passes all the same build inputs to the component packages, giving you very high chances to make use of our Cachix binary cache.
```

Reviewer 2: but it's one word, even though a compound one :thinking: maybe *build-inputs*?

Reviewer: In plain English it's not one word. If you want to use syntax, then put it in "quotes" or in `back ticks`.

```shell
nix develop github:holochain/holochain#holonix
cd my_project

> code my_file.rs

Reviewer 2: ```suggestion
codium my_file.rs
```

### `nix develop`

While working on Holochain, you will usually have an active `nix develop` to run commands. This shell overlays Holochain-specific configuration on top of your existing shell --- environment variables, Rust toolchains, binaries, libraries, and development tools --- giving you a consistent development environment to build Holochain apps. The shell environment is only set up in the current shell session, and will be reset automatically when you `exit` the shell.

> If you want to do more work, or create multiple terminals to work in, you'll need to re-enter the `nix develop` shell in each new instance. The packages are cached locally on your machine, so they will be ready the next time you enter the shell. If you follow the practices above to create a per-project `flake.nix` file, you can work completely offline, but you need an internet connection in order to enter the [ad-hoc shell](#holonix%27s-usage-of-nix%27s-flakes-features). If you're going to be working offline, you can [download a snapshot of the holochain GitHub repository](https://github.com/holochain/holochain/archive/refs/heads/develop.zip) and enter the shell once using the contained `flake.nix`. After that, the packages will be cached and you can continue to use this flake for ad-hoc shells.

Reviewer 2: re-reading this it's actually somewhat confusing. specifically, in my mind the offline instructions mentioning the snapshot and its `flake.nix` blur into setting up a custom `flake.nix`.

Author: yeah, fair enough. Is there a skeleton `flake.nix` that we can put there instead?

And also -- is it actually true that this is all you need in order to work offline? Just have a local `flake.nix` and enter it once while you're online?

Author: ```suggestion
If you want to do more work, or create multiple terminals to work in, you'll need to re-enter the `nix develop` shell in each new instance. The packages are cached locally on your machine, so they will be ready the next time you enter the shell. If you follow the practices above to create a per-project `flake.nix` file, you can work completely offline, but you need an internet connection in order to enter the [ad-hoc shell](#holonix%27s-usage-of-nix%27s-flakes-features). If you're going to be working offline, you can copy the following code into a `flake.nix` file, then enter the shell using `nix run flake.nix#holonix` once before you go offline. The packages will be all cached, and you can use the same command to enter an ad-hoc shell while you are offline.

```nix
inputs = {
    holochain-flake = {
      url = "github:holochain/holochain";
    };
};
```

just a wild guess that this would work ☝🏼 testing my understanding by writing code for review!!

Author: nope, doesn't work. The files are cached and you can enter a per-project `nix develop` shell while offline (albeit with a few spurious errors), but that skeleton didn't do anything productive.
